src/cron/syncSubscriptionWithCIO.ts

```diff
@@ -0,0 +1,38 @@
+import { syncSubscription } from '../common';
+import { StorageKey, StorageTopic, generateStorageKey } from '../config';
+import { logger } from '../logger';
+import { getRedisListLength, popFromRedisList } from '../redis';
+import { Cron } from './cron';
+
+const redisKey = generateStorageKey(
+  StorageTopic.CIO,
+  StorageKey.Reporting,
+  'global',
+);
+
+const cron: Cron = {
+  name: 'sync-subscription-with-cio',
+  handler: async (con) => {
+    logger.info('syncing subscriptions with customer.io');
+
+    while ((await getRedisListLength(redisKey)) > 0) {
+      try {
+        const userIdsRaw = await popFromRedisList(redisKey, 100);
+        if (!userIdsRaw) {
+          break;
+        }
+        // Store it in a set to remove duplicates
+        const userIds = new Set(userIdsRaw);
+
+        await syncSubscription([...userIds], con);
+
+        logger.info(`synced subscriptions for ${userIds.size} users`);
+      } catch (err) {
+        logger.error({ err }, 'error syncing subscriptions');
+        break;
+      }
```

Comment:
The error thrown would break the loop either way so it more fixes the logging part but if for any reason redis keys don't get popped properly in `popFromRedisList` we would still get infinite loop. 

Is there a way to get the keys before hand and then execute in a for loop?

Or just limit amount of keys we process in single cron (since it can only take 5 minutes), for example 1000 and solve it like that.

Reply:
Cron jobs have a max runtime of 5 minutes, but I've set it do 4 minutes for this. So it will not run forever.

Reply:
I don't really want to get the entries beforehand, as there is a possibility that the job gets killed (it's k8s after all) before it has executed all of the pre-fetched, which will then make the entries lost

Reply:
Another solution could be to make it into a function, and at the end, run `llen` to get length of list. And if it is >0, call the same function again. But that's just recursion loop

Reply:
Agree the redis logic is good as is, I would just add (at the end of `while` block:
```js
iterations += 1

if (iterations > X) {
    break
}
```

Like that we would parse amount of X and then the cron would pickup in next execution.

Reply:
Better to limit like that vs with execution limit since it can break mid execution eg. 4 minutes expire mid parsing a batch.

Reply:
Added limit of 100, is that enough. I think if it were to fail to pop, it would happen rather quick. So 100 iterations would be over quite quick?

Reply:
100 is also something we could easily adjust.

Also not sure how many possible users would be queued up. Right now we just get lots of 500 errors, but some of them are retries from CIO, so unsure of the actual amount.

Reply:
It's 100 * 100 users, so 10000 users, which seems ok to process in 5 min?

Reply:
In 5 minutes, yeah.

Could also add this between each execution, just to be sure we don't hit the 10req/s limit of CIO
https://github.com/dailydotdev/daily-api/pull/2021#discussion_r1661145136

Reply:
Yeah, agreed, left a comment on that comment 😆